### PR TITLE
compare courses prices with teachable

### PR DESCRIPTION
**Pull request by ibrahni**

### CODE REVIEW

File: src/main/java/com/baeldung/site/SitePage.java [hunk 1]
```diff
@@ -74,6 +72,26 @@ public WebElement findBodyElement() {
         return this.getWebDriver().findElement(By.xpath("//body"));
     }
 
+    public WebElement findById(String blockId){
```
Comment (asif-anwar) [inline src/main/java/com/baeldung/site/SitePage.java:75]: Let's change the parameter name from `blockId ` to just `id ` as it's implementation logic is generic.
And also move this method to BlogBaseDriver class.

File: src/main/java/com/baeldung/site/SitePage.java [hunk 2]
```diff
@@ -671,6 +689,15 @@ public boolean linkIdAndLinkAvailable(PurchaseLink link) {
         return element.getAttribute("href").contains(link.getLink());        
     }
 
+    public void clickOnPurchaseButton(PurchaseLink link) throws InterruptedException {
+        Thread.sleep(2000);
```
Comment (asif-anwar) [inline src/main/java/com/baeldung/site/SitePage.java:693]: Do we need to use `Thread.sleep()` here as in the next line we are waiting for 20 seconds anyway?
If we need to use it, let's use TestUtils.sleep();

File: src/test/java/com/baeldung/selenium/page/CoursePageUITest.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+            final Matcher matcher = pattern.matcher(element.getText());
+
+            if(matcher.find()) {
+                try {
+                    final String displayedPrice = matcher.group(0);
+                    logger.info("retrieving price {} from blog for {}", displayedPrice, purchaseLink.getLinkId());
+
+                    page.clickOnPurchaseButton(purchaseLink);
+                    assertTrue(page.containsTotalOnTeachable(), String.format("Couldn't find Course Total in teachable page %s ", page.getWebDriver()
```
Comment (asif-anwar) [inline src/test/java/com/baeldung/selenium/page/CoursePageUITest.java:62]: A couple of notes here, 

1. If the first link fails here, the test will end here with failure and it won't test the remaining links on the page. We should use an approach similar to `CommonUITest.givenOnTheCoursePage_whenAnaysingThePage_thenThePurchaseLinksAreSetupCorrectly()`. Same for lines #5o, 68, 74, 84.  Let me know if it makes sense.
2. In the failure message, we should log the course page, and link id as well. The point is - that if something fails, it should be easy to locate which link failed. We should be able to track it from the course page.

File: src/test/java/com/baeldung/selenium/page/CoursePageUITest.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+@ExtendWith(TestMetricsExtension.class)
+public class CoursePageUITest extends BaseUISeleniumTest {
+    private static final String PRICE_REGEX = "\\$\\d+";
+    final Pattern pattern = Pattern.compile(PRICE_REGEX, Pattern.MULTILINE);
+
+
+
+    @ParameterizedTest(name = " {displayName} - verify purchase price on {0}")
+    @MethodSource("com.baeldung.utility.TestUtils#pagesPurchaseLinksTestDataProviderForNonTeams()")
+    @Tag(GlobalConstants.TAG_DAILY)
```
Comment (asif-anwar) [inline src/test/java/com/baeldung/selenium/page/CoursePageUITest.java:40]: Let's remove the daily tag here.
Also, let's move CommonUITest.givenOnTheCoursePage_whenAnaysingThePage_thenThePurchaseLinksAreSetupCorrectly() in this class, and remove daily tag. Both are course related. It makes sense to have them in the same class.

**Review by asif-anwar — COMMENTED** [review]
I have left a few comments.

File: src/test/java/com/baeldung/selenium/page/CoursePageUITest.java [hunk 5]
```diff
@@ -27,61 +27,105 @@
...
                     logger.error(ex.getMessage());
                 }
-
-            }else{
-                errorRetrievingFromBlog.add(purchaseLink);
             }
         });
 
-        assertTrue(errorRetrievingFromBlog.isEmpty(), String.format(" Couldn't retrieve price for the following items [%s]", errorRetrievingFromBlog.stream().map(CoursePurchaseLinksVO.PurchaseLink::getLinkId).collect(Collectors.joining(","))));
+        assertAll(purchaseLinkDisplayedOnBlog);
```
Comment (asif-anwar) [inline src/test/java/com/baeldung/selenium/page/CoursePageUITest.java:123]: Considering the fact that if the first assertAll() fails, it won't execute the next assertAll(), my thinking is that if the price doesn't exist, we can consider it 0, we don't need to assert whether it exists. And in the final assertion, we can check whether they match and are not 0. This way, we can remove the below assertions,
assertAll(priceExistsOnBlog);
assertAll(priceDisplayedOnTeachable);
assertAll(priceExistsOnTeachable); 

Let me know what do you think?

Comment (ibrahni) [reply src/test/java/com/baeldung/selenium/page/CoursePageUITest.java:123]: Yes, we can do such a thing, but what if we couldn't retrieve the price from the teachable, should be set to 0 as well and do the comparison and we will get that the prices are equals, which is not the case.

Comment (asif-anwar) [reply src/test/java/com/baeldung/selenium/page/CoursePageUITest.java:124]: @ibrahni 
We should assert that the prices are not 0 and they match (two conditions in a single assertion). However, we should log an error message (red-colored) that prices for xx link on xx page not found. 
Let me know if it makes sense.

Comment (ibrahni) [reply src/test/java/com/baeldung/selenium/page/CoursePageUITest.java]: @asif-anwar 
Changes done, so we asserting prices the right way

**Pull request merged**